runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/jobsubmission/JobService.java

```diff
@@ -0,0 +1,309 @@
+      LOG.trace("{} {}", CancelJobRequest.class.getSimpleName(), request);
+      String invocationId = request.getJobId();
+      JobInvocation invocation = invocations.get(invocationId);
+
+      JobState.Enum state;
+      synchronized (invocation) {
+        invocation.cancel();
+        state = invocation.getState();
+      }
+
+      CancelJobResponse response = CancelJobResponse.newBuilder().setState(state).build();
+      responseObserver.onNext(response);
+      responseObserver.onCompleted();
+    } catch (Exception e) {
+      LOG.error("Encountered Unexpected Exception", e);
+      responseObserver.onError(Status.INTERNAL.withCause(e).asException());
+    }
+  }
+
+  @Override
+  public void getStateStream(
+      GetJobStateRequest request,
+      StreamObserver<GetJobStateResponse> responseObserver) {
+    try {
+      String invocationId = request.getJobId();
+      JobInvocation invocation = invocations.get(invocationId);
+
+      Function<JobState.Enum, GetJobStateResponse> responseFunction =
+          state -> GetJobStateResponse.newBuilder().setState(state).build();
+      TransformStreamObserver<JobState.Enum, GetJobStateResponse> stateObserver =
+          TransformStreamObserver.create(responseFunction, responseObserver);
+      synchronized (invocation) {
+        invocation.addStateObserver(stateObserver);
+      }
+    } catch (Exception e) {
+      LOG.error("Encountered Unexpected Exception", e);
+      responseObserver.onError(Status.INTERNAL.withCause(e).asException());
+    }
+  }
+
+  @Override
+  public void getMessageStream(
+      JobMessagesRequest request,
+      StreamObserver<JobMessagesResponse> responseObserver) {
+    try {
+      String invocationId = request.getJobId();
+      JobInvocation invocation = invocations.get(invocationId);
+
+      Function<JobState.Enum, JobMessagesResponse> stateResponseFunction =
+          state ->
+              JobMessagesResponse
+                  .newBuilder()
+                  .setStateResponse(GetJobStateResponse.newBuilder().setState(state).build())
+                  .build();
+      TransformStreamObserver<JobState.Enum, JobMessagesResponse> stateObserver =
+          TransformStreamObserver.create(stateResponseFunction, responseObserver);
+
+      Function<JobMessage, JobMessagesResponse> messagesResponseFunction =
+          message -> JobMessagesResponse.newBuilder().setMessageResponse(message).build();
+      TransformStreamObserver<JobMessage, JobMessagesResponse> messageObserver =
+          TransformStreamObserver.create(messagesResponseFunction, responseObserver);
+
+      synchronized (invocation) {
+        invocation.addStateObserver(stateObserver);
+        invocation.addMessageObserver(messageObserver);
+      }
+    } catch (Exception e) {
+      LOG.error("Encountered Unexpected Exception", e);
+      responseObserver.onError(Status.INTERNAL.withCause(e).asException());
+    }
+  }
+
+  @Override
+  public void close() throws Exception {
```

Comment:
Does anyone think we should also cancel managed job invocations here and wait for them to close?  I'm sort of torn on the issue.

Reply:
We can avoid the issue by throwing an exception if they haven't been completed yet :) Meaning, you can close the JobService only if you're done with the jobs.